### PR TITLE
Add initial support for linaria for static CSS interop

### PR DESCRIPTION
**Pull request by NateRadebaugh**
Related discussion:
* https://github.com/ben-rogerson/twin.macro/discussions/244

### CODE REVIEW

**Comment by ben-rogerson** [comment]
thanks for the PR's, you got a good base ready this feature - 🤞 it can be done!

**Comment by ben-rogerson** [comment]
I have given this a quick initial test with the `styled` import (and by including the fix from #256).

This is working:

```js
import tw from 'twin.macro'
import { styled } from 'linaria/react'

const Button = styled.button`
  ${tw`bg-black`}
`
```

But importing from twin doesn't work:

```js
import tw, { styled } from 'twin.macro'

const Button = styled.button`
  ${tw`bg-black`}
`
```

Results in `Using the "styled" tag in runtime is not supported. Make sure you have set up the Babel plugin correctly`.

## Array styling

There's also a common syntax twin uses/advertises which is adding the items to an array.
Either linaria needs some adjustments or twin (or a bit of both) to allow this:

```js
import tw from "twin.macro";

const Button = tw.button`bg-black`;

// ↓ ↓ ↓ ↓ ↓ ↓

import { styled as _styled } from "linaria/react";

const Button = _styled.button({
  "--tw-bg-opacity": "1",
  "backgroundColor": "rgba(0, 0, 0, var(--tw-bg-opacity))"
});
```

and also this:

```js
import tw, { styled } from "twin.macro";

const Button = styled.button([tw`bg-black`]);

// ↓ ↓ ↓ ↓ ↓ ↓

import { styled as _styled } from "linaria/react";

const Button = _styled.button([{
  "--tw-bg-opacity": "1",
  "backgroundColor": "rgba(0, 0, 0, var(--tw-bg-opacity))"
}]);
```

**Comment by NateRadebaugh** [comment]
Agreed, I haven't figured out the glue here but while I've got the example rendering linaria fully can't see any of the twin.macro ones working as expected. I don't understand the twin code well enough to determine the fix at the moment.

**Comment by ben-rogerson** [comment]
I'm keen to explore this further, but I'll have to circle back around and tackle it 👍

**Comment by vinverdy** [comment]
will it merged in near future?

**Comment by zaguiini** [comment]
What is this PR missing? Can I help in any way?

**Comment by NateRadebaugh** [comment]
This PR was never completed and is still just a draft. Was hoping @ben-rogerson  would pick it up or give some pointers.

**Comment by ben-rogerson** [comment]
I'm planning to wait for version 3 to come out of beta before pursuing an integration. Unless the API is similar enough?

**Comment by lukesmurray** [comment]
> I'm planning to wait for version 3 to come out of beta before pursuing an integration. Unless the API is similar enough?

I would be so excited to see twin.macro add support for a zero-runtime CSS library. With [linaria V4](https://github.com/callstack/linaria/releases/tag/v4.0.0) recently being released, I would love to help twin.macro implement support for linaria. @ben-rogerson, would you be open to PRs adding linaria support? Or is that something you would prefer to tackle on your own?

**Comment by ben-rogerson** [comment]
Hey Luke - yeah definitely open to a PR - how about a new "Linaria + Emotion + TypeScript" example?

As there's so many changes in the next `twin.macro@rc` I wouldn't advise working with any code in this repository at the moment but see how you go adding an example in the [twin.examples repo](https://github.com/ben-rogerson/twin.examples) - if you run into any issues either contact me on Discord or post up the unfinished PR and ask questions there - thanks!

**Comment by zaguiini** [comment]
Note to whoever implements this following @ben-rogerson's comment about custom imports: global styles in Linaria work a little bit differently: https://github.com/callstack/linaria/blob/v4.0.0/docs/BASICS.md#adding-global-styles.

**Comment by ben-rogerson** [comment]
With global styles we should use the `globalStyles` (style object) import rather than `GlobalStyles` (for use in jsx) import. 

Something like this:

```js
import { globalStyles } from 'twin.macro'

export const globals = css(globalStyles)
```

**Comment by lukesmurray** [comment]
sounds good i'll put one together :)

**Comment by lukesmurray** [comment]
Started working on this but ran into an issue with vite and linaria.
I submitted an issue to linaria https://github.com/callstack/linaria/issues/1044.

**Comment by lukesmurray** [comment]
Ok I gave this a first shot in a [public repo](https://github.com/lukesmurray/vite-linaria-typescript) ([codesandbox demo](https://codesandbox.io/s/mystifying-waterfall-hie59p)).

**Things that work**

- global styles
- css import
- Styled import
- interpolating arrays in css and styled imports

```tsx
// array interpolation works 👍 
css`${[tw`bg-red-500`, tw`text-sm`]}`
```

**Things that don't work**

- css prop
- tw prop
- using tw import to create styled objects

**Potential issues**

Linaria styled tags allow for conditional styling via prop interpolation functions.
However, these functions only let you conditionally determine a css _property value_, not a css _property name and property value_.
This limitation is due to the fact that linaria compiles prop interpolations into css variables and you can't represent a property name and value with a css variable.

```tsx
// this works
const Title = styled.h1`
  color: ${props => (props.primary ? 'tomato' : 'black')};
`;

// this will never work
const Title = styled.h1`
  ${props => (props.primary ? tw`text-red-500` :  tw`text-black`)};
`;
```

Linaria css interpolation is not linked to components and doesn't have access to state or props. If you want to condtionally apply styles you need to separate each potential style into a separate `css` call and conditionally apply the class names using `cx`.

```tsx
// this works
const Title = ({primary, children}) => {
  return <h1 className={cx(primary ? css`${tw`text-red-500`}` :  css`${tw`text-black`}`)}>{children}</h1>
}

// this will never work
const Title = ({primary, children}) => {
  return <h1 className={css`${primary ? tw`text-red-500` :  tw`text-black`}`}>{children}</h1>
}
```

So in summary I'm optimistic that we can get this working but we probably need to do some leg work to get the css prop working. Most likely we want the css prop to accept any of `tw`, `css` or an array of either. `tw` should probably be compiled into `css` when used at the top level and can be left alone when used inside of `css`. Arrays passed to the css prop should be merged with `cx`.

**Comment by lukesmurray** [comment]
if it's possible to implement support for the tw and css props without modifying tw.macro itself i could try to do that but i would need a little hint.

**Comment by ben-rogerson** [comment]
There’s some great findings here Luke!
You're right, further updates would need to be made in the twin.macro repo.

Let's put this on hold for a little while until I'm through the v3 updates - both me and the repo will be better primed to pick this up again and complete the new `linaria` preset.
